### PR TITLE
Apps: clean up x509 extension generation and cert signing

### PR DESCRIPTION
**Pull request by DDvO**
In preparation of a solution for #13603 I found that the code preparing X.509 signing and handling specifications for X.509 extensions should be cleaned up.
Otherwise it would add to the mess that already has started there.

I do this as a separate up-front PR in order to make reviewing easier.

* `apps/x509.c`: Factor out common aspects of X509 signing
* `apps/{req,x509,ca}.c`: Cleanup: move shared `X509{,_REQ,_CRL}` code to `apps/lib/apps.c`
* `apps/{req,x509,ca}.c`: Clean up setting X.509 cert version v3
* `x509v3_config.pod`: Clarify the semantics of `subjectKeyIdentifier` and `authorityKeyIdentifier`
* `apps/req.c`: Improve diagnostics on duplicate `-reqext` keys etc.
* `apps{ca,req,x509}.c: Improve diagnostics and doc, in particular on X.509 extensions
* `openssl-ca.pod.in`: Clarify the `-extensions`/`-crlexts` option vs. `x509_extensions`/`crl_extensions`
* `openssl_hexstr2buf_sep()`: Prevent misleading 'malloc failure' errors on short input
* `v2i_AUTHORITY_KEYID()`: Correct out-of-memory behavior and avoid mem leaks
* `X509V3_EXT_add_nconf_sk()`: Improve description and use of `sk` arg, which may be NULL

### CODE REVIEW

File: apps/req.c [hunk 1]
```diff
@@ -1708,14 +1705,25 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
...
+    int rv = 0;
 
-    if (do_sign_init(mctx, pkey, md, sigopts) > 0)
-        rv = (X509_sign_ctx(x, mctx) > 0);
+    if (sk_X509_EXTENSION_num(exts /* may be NULL */) > 0) {
+        /* Prevent X509_V_ERR_EXTENSIONS_REQUIRE_VERSION_3 */
+        if (!X509_set_version(cert, 2)) /* Make sure cert is X509 v3 */
+            goto end;
+
+        /* TODO further measures for ensuring RFC 5280 compliance */
```
Comment (t8m) [inline apps/req.c]: I am not quite sure what do you mean by this. I do not think we should disallow making RFC 5280 non-compliant certificates. If you mean adding the SKID and AKID, then OK. Otherwise anything else should be probably just a warning if anything.

Comment (DDvO) [reply apps/req.c]: Yeah, the addition of SKID and AKID (unless explicitly disabled) is what I meant primarily.
Likely there are more such things.

Generally, I'd say those apps should produce compliant certs by default.
Of course, at library level there is full freedom anyway to do weird things :wink:

Comment (DDvO) [reply apps/req.c]: I've meanwhile re-phrased this to:
```
        /* any further measures for ensuring default RFC 5280 compliance? */
```

**Comment by DDvO** [comment]
It turns out there was more stuff to clean up with the implementation and documentation of the `ca`, `req`, and `x509` extension options.

File: apps/x509.c [hunk 2]
```diff
@@ -965,7 +960,7 @@ static ASN1_INTEGER *x509_load_serial(const char *CAfile,
         goto end;
 
     if (!BN_add_word(serial, 1)) {
-        BIO_printf(bio_err, "add_word failure\n");
+        BIO_printf(bio_err, "Add_word failure\n");
```
Comment (t8m) [inline apps/x509.c:963]: Serial number increment failure

File: apps/req.c [hunk 3]
```diff
@@ -769,21 +769,21 @@ int req_main(int argc, char **argv)
 
             /* Set up V3 context struct */
 
-            X509V3_set_ctx(&ext_ctx, x509ss, x509ss, NULL, NULL, 0);
+            X509V3_set_ctx(&ext_ctx, x509ss, x509ss, NULL, NULL, X509V3_CTX_REPLACE);
```
Comment (t8m) [inline apps/req.c:772]: Why this change is needed? Should that be somehow reflected in the commit message? Similarly down below where you also add this flag.

File: doc/man5/x509v3_config.pod [hunk 4]
```diff
@@ -22,6 +23,9 @@ Each entry in the extension section takes the form:
 
 If B<critical> is present then the extension will be marked as critical.
 
+If multiple entries are processed for the same extension name,
+later entries override earlier ones with the same name.
```
Comment (t8m) [inline doc/man5/x509v3_config.pod:27]: Ah, now I get it. However this seems to me as substantial change in the processing. Perhaps at least CHANGES.md entry should be added?

File: apps/x509.c (hunk 2 above)
Comment (DDvO) [reply apps/x509.c]: Good point, done.

File: apps/req.c (hunk 3 above)
Comment (DDvO) [reply apps/req.c:772]: See below...

File: doc/man5/x509v3_config.pod (hunk 4 above)
Comment (DDvO) [reply doc/man5/x509v3_config.pod:27]: This is basically just a clarification in the code (at points like the one referred to above)
and rather prominently at this point in the documentation.
Further down (where users hardly will read it) it already said before:
```
OpenSSL does not support multiple occurrences of the same field within a
section. In this example:

 [extensions]
 subjectAltName = @alt_section

 [alt_section]
 email = steve@here
 email = steve@there

will only recognize the last value.
```
Note that this is not 'just' an OpenSSL restriction but also due to an RFC 5280 requirement.
The intro of section 4.2 says:
> A certificate MUST NOT include more than one instance of a particular extension.

So I don't really see the need for an addition to CHANGES.md on that, 
but I can do if you prefer to have one - if so, how do you suggest to phrase it?

File: apps/req.c (hunk 3 above)
Comment (DDvO) [reply apps/req.c:772]: This turned out to be helpful with the approach I plan to submit on solving #13603
because there, when a user indicates "no" for an SKID or AKID, it will tentatively add an empty entry to track this,
and just before signing it will either remove the empty entry if present and else add a default entry if nothing is present yet.

File: doc/man5/x509v3_config.pod (hunk 4 above)
Comment (t8m) [reply doc/man5/x509v3_config.pod:27]: OK, then no CHANGES.md are needed, but could you please amend the commit message to explain the change more?

Comment (DDvO) [reply doc/man5/x509v3_config.pod:27]: Good point - done.

**Comment by DDvO** [comment]
I don't understand why suddenly various CI tests fail. 
I cannot reproduce them locally and they appear to be not related to this PR.
I tried getting rid of them by rebasing on an older version of master (namely commit abdd3fa04f3401c4a542257fdd803ff4c4daf8ef), but this did not help.

**Comment by t8m** [comment]
Master should be OK again, please rebase again.

**Review by t8m — APPROVED** [review]
Approval is pending on rerun of the CI on clean rebase.

**Comment by DDvO** [comment]
> Approval is pending on rerun of the CI on clean rebase.

Thanks - after a fresh rebase the CI works fine again.

**Comment by DDvO** [comment]
Merged - thanks again @t8m for the swift handling.
